File: typescript/src/plan.ts
import { graphDigest } from "./canonical.js";
import type { EffectiveEdge, GraphPlan, JsonObject } from "./types.js";

type RecordValue = Record<string, unknown>;

function isRecord(value: unknown): value is RecordValue {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function records(value: unknown): Record<string, RecordValue> {
  if (!isRecord(value)) return {};
  return Object.fromEntries(Object.entries(value).filter((entry): entry is [string, RecordValue] => isRecord(entry[1])));
}

function strings(value: unknown): string[] {
  return Array.isArray(value) ? value.filter((item): item is string => typeof item === "string") : [];
}

export function graphEffectiveEdges(document: JsonObject): EffectiveEdge[] {
  const nodes = records(document.nodes);
  const edges: EffectiveEdge[] = [];
  for (const [nodeId, node] of Object.entries(nodes)) {
    for (const dependency of strings(node.depends_on)) edges.push({ from: dependency, to: nodeId, kind: "sequence" });
  }
  if (Array.isArray(document.edges)) {
    for (const raw of document.edges) {
      if (!isRecord(raw) || typeof raw.from !== "string" || typeof raw.to !== "string") continue;
      const edge: EffectiveEdge = { from: raw.from, to: raw.to, kind: typeof raw.kind === "string" ? raw.kind : "sequence" };
      if (typeof raw.when === "string") edge.when = raw.when;
      edges.push(edge);
    }
  }
  const seen = new Set<string>();
  return edges.filter((edge) => {
    const key = `${edge.from}\u0000${edge.to}\u0000${edge.kind}\u0000${edge.when ?? ""}`;
    if (seen.has(key)) return false;
    seen.add(key);
    return true;
  });
}

export function topologicalOrder(document: JsonObject): string[] {
  const ids = Object.keys(records(document.nodes));
  const edges = graphEffectiveEdges(document).filter((edge) => ids.includes(edge.from) && ids.includes(edge.to));
  const incoming = new Map(ids.map((id) => [id, 0]));
  const outgoing = new Map(ids.map((id) => [id, [] as string[]]));
  for (const edge of edges) {
    incoming.set(edge.to, (incoming.get(edge.to) ?? 0) + 1);
    outgoing.get(edge.from)?.push(edge.to);
  }
  const ready = ids.filter((id) => incoming.get(id) === 0).sort();
  const order: string[] = [];
  while (ready.length > 0) {
    const current = ready.shift();
    if (!current) break;
    order.push(current);
    for (const next of (outgoing.get(current) ?? []).sort()) {
      const count = (incoming.get(next) ?? 0) - 1;
      incoming.set(next, count);
      if (count === 0) {
        ready.push(next);
        ready.sort();
      }
    }
  }
  if (order.length !== ids.length) throw new Error("cannot plan a cyclic graph");
  return order;
}

function fragmentExecutions(fragment: RecordValue, named: Record<string, RecordValue>, active = new Set<string>()): number {
  return nodeExecutions(records(fragment.nodes), named, active);
}

function nodeExecutions(nodes: Record<string, RecordValue>, named: Record<string, RecordValue>, active = new Set<string>()): number {
  let total = 0;
  for (const node of Object.values(nodes)) {
    total += 1;
    const type = typeof node.type === "string" ? node.type : "task";
    if (!["loop", "map", "subgraph"].includes(type)) continue;
    const block = isRecord(node[type]) ? node[type] : {};
    let body: RecordValue | undefined;
    if (type === "subgraph" && isRecord(block.inline)) body = block.inline;
    else if ((type === "loop" || type === "map") && isRecord(block.body)) body = block.body;
    const use = typeof block.use === "string" ? block.use : undefined;
    if (!body && use && !active.has(use)) body = named[use];
    if (!body) continue;
    const nested = fragmentExecutions(body, named, use ? new Set(active).add(use) : active);
    if (type === "loop") total += nested * (typeof block.max_iterations === "number" ? block.max_iterations : 1);
    else if (type === "map") total += nested * (typeof block.max_items === "number" ? block.max_items : 1);
    else total += nested;
  }
  return total;
}

export function planGraph(document: JsonObject): GraphPlan {
  const nodes = records(document.nodes);
  const edges = graphEffectiveEdges(document);
  const order = topologicalOrder(document);
  const entrypoints = strings(document.entrypoints);
  const outgoing = new Map(Object.keys(nodes).map((id) => [id, [] as string[]]));
  for (const edge of edges) outgoing.get(edge.from)?.push(edge.to);
  const reachableSet = new Set<string>();
  const stack = [...entrypoints];
  while (stack.length > 0) {
    const current = stack.pop();
    if (!current || reachableSet.has(current)) continue;
    reachableSet.add(current);
    stack.push(...(outgoing.get(current) ?? []));
  }
  const tierHistogram: Record<string, number> = {};
  const unsupported = new Set<string>();
  for (const node of Object.values(nodes)) {
    const intelligence = isRecord(node.intelligence) ? node.intelligence : {};
    const tier = typeof intelligence.tier === "string" ? intelligence.tier : "unspecified";
    tierHistogram[tier] = (tierHistogram[tier] ?? 0) + 1;
    const type = typeof node.type === "string" ? node.type : "task";
    if (type !== "task" && type !== "gate") unsupported.add(type);
  }
  return {
    graphId: typeof document.id === "string" ? document.id : "",
    graphDigest: graphDigest(document),
    order,
    entrypoints,
    effectiveEdges: edges,
    reachable: order.filter((id) => reachableSet.has(id)),
    unreachable: order.filter((id) => !reachableSet.has(id)),
    tierHistogram,
    worstCaseNodeExecutions: nodeExecutions(nodes, records(document.subgraphs)),
    executable: false,
    unsupportedFeatures: [...unsupported].sort(),
  };
}

File: rust/src/plan.rs
use std::collections::{BTreeMap, BTreeSet, VecDeque};

use thiserror::Error;

use crate::{Document, EffectiveEdge, GraphPlan, graph_digest, object, strings};

/// Failure while constructing a deterministic graph plan.
#[derive(Debug, Error)]
pub enum PlanError {
    /// The effective graph contains a directed cycle.
    #[error("graph contains a cycle")]
    Cycle,
    /// Canonical graph identity calculation failed.
    #[error(transparent)]
    Digest(#[from] crate::canonical::CanonicalError),
}

/// Combines `depends_on` relationships and explicit edges into a sorted edge set.
pub fn graph_effective_edges(document: &Document) -> Vec<EffectiveEdge> {
    let nodes = object(document.get("nodes"));
    let mut edges = Vec::new();
    let mut seen = BTreeSet::new();
    for (id, raw) in nodes {
        for dependency in strings(raw.as_object().and_then(|node| node.get("depends_on"))) {
            let key = (dependency.clone(), id.clone(), "sequence".to_owned());
            if seen.insert(key.clone()) {
                edges.push(EffectiveEdge {
                    from: key.0,
                    to: key.1,
                    kind: key.2,
                    when: None,
                });
            }
        }
    }
    for raw in document
        .get("edges")
        .and_then(|value| value.as_array())
        .into_iter()
        .flatten()
    {
        let edge = object(Some(raw));
        let from = edge
            .get("from")
            .and_then(|v| v.as_str())
            .unwrap_or_default()
            .to_owned();
        let to = edge
            .get("to")
            .and_then(|v| v.as_str())
            .unwrap_or_default()
            .to_owned();
        let kind = edge
            .get("kind")
            .and_then(|v| v.as_str())
            .unwrap_or("sequence")
            .to_owned();
        if seen.insert((from.clone(), to.clone(), kind.clone())) {
            edges.push(EffectiveEdge {
                from,
                to,
                kind,
                when: edge.get("when").and_then(|v| v.as_str()).map(str::to_owned),
            });
        }
    }
    edges.sort_by(|left, right| {
        (&left.from, &left.to, &left.kind).cmp(&(&right.from, &right.to, &right.kind))
    });
    edges
}

/// Computes a deterministic, identifier-tiebroken topological node order.
pub fn topological_order(document: &Document) -> Result<Vec<String>, PlanError> {
    let nodes = object(document.get("nodes"));
    let edges = graph_effective_edges(document);
    let mut incoming: BTreeMap<String, usize> = nodes.keys().map(|id| (id.clone(), 0)).collect();
    let mut outgoing: BTreeMap<String, Vec<String>> =
        nodes.keys().map(|id| (id.clone(), vec![])).collect();
    for edge in &edges {
        if incoming.contains_key(&edge.from) && incoming.contains_key(&edge.to) {
            *incoming.get_mut(&edge.to).expect("known node") += 1;
            outgoing
                .get_mut(&edge.from)
                .expect("known node")
                .push(edge.to.clone());
        }
    }
    for targets in outgoing.values_mut() {
        targets.sort();
    }
    let mut ready: BTreeSet<String> = incoming
        .iter()
        .filter(|(_, count)| **count == 0)
        .map(|(id, _)| id.clone())
        .collect();
    let mut order = Vec::with_capacity(nodes.len());
    while let Some(id) = ready.pop_first() {
        order.push(id.clone());
        for target in outgoing.get(&id).into_iter().flatten() {
            let count = incoming.get_mut(target).expect("known target");
            *count -= 1;
            if *count == 0 {
                ready.insert(target.clone());
            }
        }
    }
    if order.len() != nodes.len() {
        return Err(PlanError::Cycle);
    }
    Ok(order)
}

/// Produces a deterministic, non-executing AGS conformance-level-0 plan.
pub fn plan_graph(document: &Document) -> Result<GraphPlan, PlanError> {
    let nodes = object(document.get("nodes"));
    let edges = graph_effective_edges(document);
    let order = topological_order(document)?;
    let entrypoints = strings(document.get("entrypoints"));
    let mut outgoing: BTreeMap<String, Vec<String>> = BTreeMap::new();
    for edge in &edges {
        outgoing
            .entry(edge.from.clone())
            .or_default()
            .push(edge.to.clone());
    }
    let mut reachable = BTreeSet::new();
    let mut queue: VecDeque<String> = entrypoints.iter().cloned().collect();
    while let Some(id) = queue.pop_front() {
        if reachable.insert(id.clone()) {
            queue.extend(outgoing.get(&id).into_iter().flatten().cloned());
        }
    }
    let unreachable = nodes
        .keys()
        .filter(|id| !reachable.contains(*id))
        .cloned()
        .collect();
    let mut histogram = BTreeMap::new();
    let mut worst_case = 0_u64;
    let mut unsupported = BTreeSet::new();
    for raw in nodes.values() {
        let node = object(Some(raw));
        let tier = object(node.get("intelligence"))
            .get("tier")
            .and_then(|v| v.as_str())
            .unwrap_or("none");
        *histogram.entry(tier.to_owned()).or_insert(0) += 1;
        let mut executions = 1_u64;
        if node.get("type").and_then(|v| v.as_str()) == Some("loop") {
            executions = object(node.get("loop"))
                .get("max_iterations")
                .and_then(|v| v.as_u64())
                .unwrap_or(1);
            unsupported.insert("loop".to_owned());
        }
        if node.get("type").and_then(|v| v.as_str()) == Some("map") {
            executions = object(node.get("map"))
                .get("max_items")
                .and_then(|v| v.as_u64())
                .unwrap_or(1);
            unsupported.insert("map".to_owned());
        }
        if matches!(
            node.get("type").and_then(|v| v.as_str()),
            Some("decision" | "subgraph")
        ) {
            unsupported.insert(
                node.get("type")
                    .and_then(|v| v.as_str())
                    .unwrap()
                    .to_owned(),
            );
        }
        worst_case = worst_case.saturating_add(executions);
    }
    Ok(GraphPlan {
        graph_id: document
            .get("id")
            .and_then(|v| v.as_str())
            .unwrap_or_default()
            .to_owned(),
        graph_digest: graph_digest(document)?,
        order,
        entrypoints,
        effective_edges: edges,
        reachable: reachable.into_iter().collect(),
        unreachable,
        tier_histogram: histogram,
        worst_case_node_executions: worst_case,
        executable: false,
        unsupported_features: unsupported.into_iter().collect(),
    })
}
